### PR TITLE
whisper.py: numpy alternative for unpack & format

### PR DESCRIPTION
**Pull request by gcoudu**
Benchmark per number of points fetched, in seconds:
+-------------+---------+----------+-------+
|   Points    |  Numpy  |  Struct  | Ratio |
+-------------+---------+----------+-------+
|  1          | 0.00326 | 0.0023   | -29%  |
|  100        | 0.00353 | 0.00475  | 35%   |
|  10,000     | 0.01675 | 0.18555  | 1008% |
|  1,000,000  | 1.64262 | 29.75022 | 1711% |
+-------------+---------+----------+-------+

The numpy unpacking and formatting option is faster than the struct option when fetching more than 50 points. This method speeds up exponentially as the number of points increases but is slower of 0.001 second when fetching 1 single point.

We also noticed that whisper does not release memory when used via a daemon service. This led to critical memory issues on our side.
After investigating this issue, we attributed it to the struct.unpack automatic caching.
Manual cache clearing after each request partially solved the problem but at the cost of huge CPU usage.

This numpy option fixes this issue, releasing all memory after request with minimal CPU usage.
On top of that, the values are returned as a numpy array allowing easy computation transformation and can be jsonified through .tolist()

We use it at flockpit.com. Thank you for your amazing work on Graphite!

### CODE REVIEW

**Comment by deniszh** [comment]
Hi @gcoudu,
Interesting patch, thanks a lot for sharing. Could be useful if we make numpy completely optional dependency.
Did you use your patched whisper with carbon/graphite or in your own project?

**Comment by gcoudu** [comment]
Hi @deniszh,
Thanks for your comment. We use statsd/carbon for the aggregation and then fetch directly through our patched whisper.

We apply our transformations directly on the numpy array (std, moving average, aggregation per interval, etc...) and convert it via .tolist() to return it as json.

As we are working in the financial industry, we are dealing mainly with minute data and on long timeframes, hence this numpy approach drastically improved our latency.

But the real improvement came for the memory consumption. The unpatched whisper was consuming more and more memory and had to be rebooted off and on. This patched version made our whole infrastructure more stable and scalable.

File: whisper.py [hunk 1]
```diff
@@ -32,6 +32,7 @@
 import struct
 import sys
 import time
+import numpy
```
Comment (deniszh) [inline whisper.py]: Conditional import will be better here, we do not want to bring numpy as default dependency. Like
```python
try:
    import numpy
except ImportError as err:
    pass
```

File: whisper.py [hunk 2]
```diff
@@ -745,7 +751,7 @@ def info(path):
   return None
 
 
-def fetch(path, fromTime, untilTime=None, now=None):
+def fetch(path, fromTime, untilTime=None, now=None, asNumpy=True):
```
Comment (deniszh) [inline whisper.py]: Please change default to ```asNumpy=False```

File: whisper.py [hunk 3]
```diff
@@ -758,10 +764,10 @@ def fetch(path, fromTime, untilTime=None, now=None):
 Returns None if no data can be returned
 """
   with open(path, 'rb') as fh:
-    return file_fetch(fh, fromTime, untilTime, now)
+    return file_fetch(fh, fromTime, untilTime, now, asNumpy)
 
 
-def file_fetch(fh, fromTime, untilTime, now=None):
+def file_fetch(fh, fromTime, untilTime, now=None, asNumpy=True):
```
Comment (deniszh) [inline whisper.py]: Please change default to ```asNumpy=False```

File: whisper.py [hunk 4]
```diff
@@ -795,10 +801,10 @@ def file_fetch(fh, fromTime, untilTime, now=None):
     if archive['retention'] >= diff:
       break
 
-  return __archive_fetch(fh, archive, fromTime, untilTime)
+  return __archive_fetch(fh, archive, fromTime, untilTime, asNumpy)
 
 
-def __archive_fetch(fh, archive, fromTime, untilTime):
+def __archive_fetch(fh, archive, fromTime, untilTime, asNumpy=True):
```
Comment (deniszh) [inline whisper.py]: Please change default to ```asNumpy=False```

File: whisper.py [hunk 5]
```diff
@@ -843,25 +849,34 @@ def __archive_fetch(fh, archive, fromTime, untilTime):
...
   step = archive['secondsPerPoint']
+  timeInfo = (fromInterval, untilInterval, step)
 
-  for i in xrange(0, len(unpackedSeries), 2):
-    pointTime = unpackedSeries[i]
-    if pointTime == currentInterval:
-      pointValue = unpackedSeries[i+1]
-      valueList[i//2] = pointValue  # In-place reassignment is faster than append()
-    currentInterval += step
+  if asNumpy:
```
Comment (deniszh) [inline whisper.py]: Skip NumPy implementation if it's not loaded
```python
if asNumpy and pointFormatNumpy:
```

**Comment by deniszh** [comment]
I'm just wondering is Graphite could gain some profit from that patch. But of course, whisper is separate library, so, your patch is valuable by its own. Just add some comments to make your code more tolerant to missing numpy, hope it will help with tests...

**Comment by deniszh** [comment]
Closed in favour of https://github.com/graphite-project/whisper/pull/143

**Pull request closed without merging**